torch/nn/functional.py

```diff
@@ -2258,6 +2258,122 @@ def mse_loss(input, target, size_average=None, reduce=None, reduction='mean'):
     return ret
 
 
+def _fspecial_gaussian(size, channel, sigma):
+    coords = torch.tensor([(x - (size - 1.) / 2.) for x in range(size)])
+    coords = -coords ** 2 / (2. * sigma ** 2)
+    grid = coords.view(1, -1) + coords.view(-1, 1)
+    grid = grid.view(1, -1)
+    grid = grid.softmax(-1)
+    kernel = grid.view(1, 1, size, size)
+    kernel = kernel.expand(channel, 1, size, size).contiguous()
+    return kernel
+
+
+def _ssim(input, target, max_val, k1, k2, channel, kernel):
+    c1 = (k1 * max_val) ** 2
+    c2 = (k2 * max_val) ** 2
+
+    mu1 = conv2d(input, kernel, groups=channel)
+    mu2 = conv2d(target, kernel, groups=channel)
+
+    mu1_sq = mu1 ** 2
+    mu2_sq = mu2 ** 2
+    mu1_mu2 = mu1 * mu2
+
+    sigma1_sq = conv2d(input * input, kernel, groups=channel) - mu1_sq
+    sigma2_sq = conv2d(target * target, kernel, groups=channel) - mu2_sq
+    sigma12 = conv2d(input * target, kernel, groups=channel) - mu1_mu2
+
+    v1 = 2 * sigma12 + c2
+    v2 = sigma1_sq + sigma2_sq + c2
+
+    ssim = ((2 * mu1_mu2 + c1) * v1) / ((mu1_sq + mu2_sq + c1) * v2)
+    return ssim, v1 / v2
+
+
+@weak_script
+def ssim_loss(input, target, max_val, filter_size=11, k1=0.01, k2=0.03,
+              sigma=1.5, size_average=None, reduce=None, reduction='mean'):
+    # type: (Tensor, Tensor, float, int, float, float, float, Optional[bool], Optional[bool], str) -> Tensor
+    r"""ssim_loss(input, target, max_val, filter_size, k1, k2,
+                  sigma, size_average=None, reduce=None, reduction='mean') -> Tensor
+
+    Measures the structural similarity index (SSIM) error.
+
+    See :class:`~torch.nn.SSIMLoss` for details.
+    """
+
+    dim = input.dim()
+    if dim != 4:
+        raise ValueError('Expected 4 dimensions (got {})'.format(dim))
+
+    if input.size() != target.size():
+        raise ValueError('Expected input size ({}) to match target size ({}).'
+                         .format(input.size(0), target.size(0)))
+
+    if size_average is not None or reduce is not None:
+        reduction = _Reduction.legacy_get_string(size_average, reduce)
+
+    _, channel, _, _ = input.size()
+    kernel = _fspecial_gaussian(filter_size, channel, sigma)
```

Comment:
can these filters please be cached in the module version?

Reply:
> can these filters please be cached in the module version?

sure, I have moved the `kernel` in the module

Reply:
Will caching the kernel work when multiple GPUs are being used? The data transfer overhead can be minimized if the kernel and weights are initialized on the device (GPU) instead of being initialized on host (CPU) and then transferred to device. However, I do not know what Pytorch sends as memory copies and what it uses as instructions.

Reply:
@veritas9872 If we can get the specific device (`cuda` or `cpu`) that `nn.SSIMLoss` or `nn.MSSSIMLoss` are initialized on, we can initialize the kernel on the specific device instead of on `cpu` that transfer it to other devices. However, I cannot find any to detect the used device during initialize Loss function class.  Do you have any advice?

Reply:
> Will caching the kernel work when multiple GPUs are being used?

I think `kernel = kernel.to(device=input.device)` can move the tensor `kernel` into multiple GPUs according to the used device of tensor `input`.

Reply:
But won't this result in the kernel being moved from one device to another every time that a different device is used? 
I think it would be best if there were a copy of the kernel and weights in every device.

Reply:
I am not very familiar with how nn.DataParallel and nn.DistributedDataParallel work, but is there no way to use them to make multiple copies?

Reply:
@veritas9872  I think nn.DataParallel is only applied on models instead of tensor itself. So, we can get complete device list of tensor when calculating loss just by calling tensor.device. Looking for explanations from experts 😂

Reply:
Can we perhaps register the kernel and weights as buffers? All losses are subclasses of Pytorch modules, so they can have buffers added to them. See [here](https://pytorch.org/docs/stable/nn.html#torch.nn.Module.register_buffer) for details.

Reply:
DataParallel and DistributedDataParallel might have native buffer copying. I am not sure but other Modules have parameters copied without any problems.

Reply:
I looked it up but this would require the loss function to be a DataParallel model as well (I think). It might be more flexible to generate the kernel every time, but make it so that it is generated on the device (GPU) that it is being used on, while minimizing the number and amount of data that has to passed from host (CPU) to device every time that the function has to be calculated.

Reply:
BCEWithLogitsLoss also has weights that it needs to store and it seems to use buffers. So there is a possibility that this will solve the problem of caching for multiple devices. However, I don't know how the inner mechanics of Module buffers work well enough to know whether this will work for SSIM.

Reply:
@veritas9872  After reading this post: https://discuss.pytorch.org/t/how-pytorchs-parallel-method-and-distributed-method-works/30349/2 . I think Pytorch calculate the loss on default GPU:0 or in CPU. So, we really don't need to cache the kernel on multiple devices. Neither current implementation will lead to 
> kernel being moved from one device to another every time

because the tensor just uses the same and signle device.